tests/e2e/e2e_groups_test.go

```diff
@@ -0,0 +1,254 @@
+package e2e
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/cosmos/cosmos-sdk/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	group "github.com/cosmos/cosmos-sdk/x/group"
+)
+
+type PolicyRequest struct {
+	Type      string
+	Threshold string
+	Windows   group.DecisionPolicyWindows
+}
+
+var (
+	aliceAddr                string
+	bobAddr                  string
+	charlieAddr              string
+	members                  []GroupMember
+	originalMembersFilename  = "members1.json"
+	addMemberFilename        = "members2.json"
+	removeMemberFilename     = "members3.json"
+	thresholdPolicyFilename  = "policy1.json"
+	percentagePolicyFilename = "policy2.json"
+
+	proposalMsgSendPath = "proposal1.json"
+	sendAmount          = sdk.NewInt64Coin(photonDenom, 5000000)
+	windows             = DecisionPolicyWindow{
+		MinExecutionPeriod: (0 * time.Second).String(),
+		VotingPeriod:       (30 * time.Second).String(),
+	}
+
+	thresholdPolicy = ThresholdPolicy{
+		Type:      "/cosmos.group.v1.ThresholdDecisionPolicy",
+		Threshold: "1",
+		Windows:   windows,
+	}
+
+	percentagePolicy = PercentagePolicy{
+		Type:       "/cosmos.group.v1.PercentageDecisionPolicy",
+		Percentage: "0.5",
+		Windows:    windows,
+	}
+)
+
+func (s *IntegrationTestSuite) GroupsSendMsgTest() {
+	thresholdPolicyMetadata := "Policy 1"
+	percentagePolicyMetadata := "Policy 2"
+	groupId := 1
+	proposalId := 1
+
+	var res group.QueryGroupMembersResponse
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel()
+
+	chainAAPIEndpoint := fmt.Sprintf("http://%s", s.valResources[s.chainA.id][0].GetHostPort("1317/tcp"))
+	adminAddr, err := s.chainA.validators[0].keyInfo.GetAddress()
+	s.Require().NoError(err)
+
+	val2, err := s.chainA.validators[1].keyInfo.GetAddress()
+	s.Require().NoError(err)
+	aliceAddr = val2.String()
+
+	bobAddr = s.executeGaiaKeysAddCommand(ctx, s.chainA, 0, "bob")
+	charlieAddr = s.executeGaiaKeysAddCommand(ctx, s.chainA, 0, "charlie")
+
+	s.prepareGroupFiles(ctx, s.chainA, adminAddr.String(), aliceAddr, bobAddr, charlieAddr)
```

Comment:
feels like this logic should live inside this method. it's not used by any other function and it changes the values of global variables. For instance I'd expect line 80 (`len(members)`) to be 0 based on the variable being initialized above. If the reader is just expected to jump down to the `prepareGroupFiles` definition then it hasn't actually improved legibility of the code by putting it into a separate function but actually made it harder to follow. Another way could be to make it a functional method that doesn't change state itself but returns the file and returns `members` that are then re-assigned in line.

this is low priority feedback and I wouldn't call it essential to be changed for this PR to be merged but still hopefully valuable feedback from perspective of someone reading tests without the context you built up by writing them.

Reply:
Yeah I can see how that could be confusing...outside of writing an additional function that just gets the length of the original list of members I don't know that putting that setup in the body of the main test helps with legibility but it's worth a discussion when we set aside time to do the refactor.